man/bump-license-year.1
.TH bump-license-year "1" "March 23, 2025" "" "bump-license-year manual"

.SH NAME
bump-license-year - bumps the year in LICENSE and README.md files

.SH SYNOPSIS
.B bump-license-year
[-m|--message
.I message
] [-n|--name
.I
name
] [-y|--year
.I
year
]
.br
[-d|--dry-run] [-v|verbose] [-V|--version] [-h|--help]
.RE

.SH DESCRIPTION
bump-license-year bumps the last year in the license text in LICENSE and README.md files.

.SH OPTIONS
.B
.IP "-m|--message message"
commit message
.B
.IP "-n|--name name"
owner of the license (default is git user.name)
.B
.IP "-y|--year year"
the last year (default is the current one)
.B
.IP "-d|--dry-run"
print the changes on the console only
.B
.IP "-v|--verbose"
print the changes on the console too
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE
The default commit message: "docs: Bump license year".

.SH EXAMPLES
.B
.IP "bump-license-year"
Bump the last year in the license text in LICENSE and README.md files.
.B
.IP "bump-license-year -n 'Ferdinand Prantl' -dv"
Bump the last year in the license text in LICENSE and README.md files.
Use the specific user name, print verbose messages and do not actually
make the changes - dry run.

.SH BUGS
bump-license-year's bug tracker is located at <https://github.com/prantlf/v-bump-license-year/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/v-bump-license-year#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.2.0](https://github.com/prantlf/v-bump-license-year/compare/v0.1.0...v0.2.0) (2025-03-23)

### Features

* Improve default progress output ([41ca7c3](https://github.com/prantlf/v-bump-license-year/commit/41ca7c36fe978ee825a00f8c13716f4c0257e6cd))

### Bug Fixes

* Fix man pages ([f3f232a](https://github.com/prantlf/v-bump-license-year/commit/f3f232a0e8532a30eb240d1122af183829ebff3d))

## [0.1.0](https://github.com/prantlf/v-bump-license-year/compare/v0.0.1...v0.1.0) (2025-03-22)

### Features

* Initial release ([33aa4ac](https://github.com/prantlf/v-bump-license-year/commit/33aa4acc41394f4c5d567e748dd913b14afa2d08))

## 0.0.1 (2025-03-22)

Prepare repository


src/bump-license-year.v
import os { read_lines, write_file }
import time { now }
import strconv { format_int }
import prantlf.cli { Cli, run }
import prantlf.debug { new_debug }
import prantlf.osutil { execute }
import prantlf.pcre2 { NoMatch, NoReplace, RegEx, pcre2_compile }

const version = '0.2.0'

const usage = 'Bumps the last year in the license text in LICENSE and README.md files.

Usage: bump-license-year [options]

Options:
  -m|--message <message>  commit message
  -n|--name <name>        owner of the license (default is git user.name)
  -y|--year <year>        the last year (default is the current one)
  -d|--dry-run            print the changes on the console only
  -v|--verbose            print the changes on the console too
  -V|--version            print the version of the executable and exits
  -h|--help               print the usage information and exits

The default commit message: "docs: Bump license year".

Examples:
  $ bump-license-year
  $ bump-license-year -n "Ferdinand Prantl" -dv'

struct Opts {
	message string = 'docs: Bump license year'
	name    string
	year    string
	verbose bool
	dry_run bool @[json: 'dry-run']
}

const d = new_debug('bump-license-year')

fn main() {
	run(Cli{
		usage:   usage
		version: version
	}, body)
}

fn body(mut opts Opts, _args []string) ! {
	name := if opts.name != '' {
		opts.name
	} else {
		get_name()!
	}
	year := if opts.year != '' {
		opts.year
	} else {
		get_year()
	}
	two_year_re := pcre2_compile(' (\\d+)-(\\d+) ${name}$', 0)!
	defer { two_year_re.free() }
	one_year_re := pcre2_compile(' (\\d+) ${name}', 0)!
	defer { one_year_re.free() }
	mut replaced := false
	if bump_license_year('LICENSE', name, year, two_year_re, one_year_re, opts)! {
		replaced = true
	}
	if bump_license_year('README.md', name, year, two_year_re, one_year_re, opts)! {
		replaced = true
	}
	if replaced {
		output := execute('git commit -am "${opts.message}"')!
		println(output)
		if opts.verbose {
			commit := execute('git show')!
			println(commit)
		}
	}
}

fn get_name() !string {
	name := execute('git config user.name')!
	d.log('git user.name: "%s"', name)
	return name
}

fn get_year() string {
	current_time := now()
	year := format_int(current_time.year, 10)
	d.log('current year: "%s"', year)
	return year
}

fn bump_license_year(file_name string, name string, year string, two_year_re &RegEx, one_year_re &RegEx, opts &Opts) !bool {
	dfile_name := d.rwd(file_name)
	d.log('reading "%s"', dfile_name)
	lines := read_lines(file_name)!
	d.log('received %d lines', lines.len)
	dry_run := if opts.dry_run {
		' (dry run)'
	} else {
		''
	}
	mut new_lines := []string{len: 0, cap: lines.len + 1}
	mut different := false
	mut replaced := false
	for line in lines {
		if replaced {
			new_lines << line
			continue
		}
		if new_line := two_year_re.replace(line, ' $1-${year} ${name}', pcre2.opt_replace_groups) {
			d.log('replacing "%s" with "%s"', line, new_line)
			println('Replacing ${file_name}${dry_run}: ${new_line}')
			new_lines << new_line
			different = line != new_line
			replaced = true
		} else {
			if err is NoMatch {
				if new_line := one_year_re.replace(line, ' ${year} ${name}', pcre2.opt_replace_groups) {
					d.log('replacing "%s" with "%s"', line, new_line)
					println('Replacing ${file_name}${dry_run}: ${new_line}')
					new_lines << new_line
					different = line != new_line
					replaced = true
				} else {
					if err is NoMatch {
						new_lines << line
					} else if err is NoReplace {
						d.log('no need to replace "%s"', line)
						println('Retaining ${file_name}${dry_run}: ${line}')
						new_lines << line
						different = false
						replaced = true
					} else {
						return err
					}
				}
			} else if err is NoReplace {
				d.log('no need to replace "%s"', line)
				println('Retaining ${file_name}${dry_run}: ${line}')
				new_lines << line
				different = false
				replaced = true
			} else {
				return err
			}
		}
	}
	if replaced {
		if different {
			if !opts.dry_run {
				d.log('writing "%s"', dfile_name)
				new_lines << ''
				new_text := new_lines.join('\n')
				write_file(file_name, new_text)!
				return true
			}
		}
	} else {
		println('No license line found in "${file_name}".')
	}
	return false
}
